File: src/ui/lcd/pitch-edit.ts
/**
 * LCD Pitch Edit screen — pitch pattern with step select + encoder adjust.
 * 2×8 grid layout matching the physical step button grid.
 * All text ≥16px for readability on 3.5" TFT at 50cm.
 */

import type { SequencerState } from '../../engine/types'
import type { UIState } from '../hw-types'
import { COLORS, midiToNoteName } from '../colors'
import { fillRect, strokeRect, drawText, LCD_W, LCD_CONTENT_Y, LCD_CONTENT_H } from '../renderer'

const PAD = 8
const HEADER_H = 42
const STEP_AREA_TOP = LCD_CONTENT_Y + HEADER_H
const COLS = 8
const ROW_GAP = 8
const COL_GAP = 2
const STEP_W = (LCD_W - PAD * 2 - (COLS - 1) * COL_GAP) / COLS
const AVAIL_H = LCD_CONTENT_H - HEADER_H - 4
const BAR_MAX_H = (AVAIL_H - ROW_GAP) / 2

export function renderPitchEdit(ctx: CanvasRenderingContext2D, engine: SequencerState, ui: UIState): void {
  const track = engine.tracks[ui.selectedTrack]
  const trackColor = COLORS.track[ui.selectedTrack]
  const pageOffset = ui.currentPage * 16
  const maxPage = Math.max(0, Math.ceil(track.pitch.length / 16) - 1)

  // Title
  drawText(ctx, `PITCH — T${ui.selectedTrack + 1}`, PAD, LCD_CONTENT_Y + 18, trackColor, 18)

  // Selected step info — prominent display
  const selIdx = pageOffset + ui.selectedStep
  if (selIdx < track.pitch.length) {
    const step = track.pitch.steps[selIdx]
    const slideLabel = step.slide > 0 ? `  SLD:${Math.round(step.slide * 1000)}ms` : ''
    drawText(ctx, `${midiToNoteName(step.note)} (${step.note})${slideLabel}`, PAD, LCD_CONTENT_Y + 36, COLORS.textBright, 16)
  }

  // Right side info
  let infoText = `LEN ${track.pitch.length}`
  if (track.pitch.clockDivider > 1) infoText += `  ÷${track.pitch.clockDivider}`
  if (maxPage > 0) infoText += `  P${ui.currentPage + 1}/${maxPage + 1}`
  drawText(ctx, infoText, LCD_W - PAD, LCD_CONTENT_Y + 18, COLORS.textDim, 16, 'right')

  // Use RandomConfig pitch range for stable normalization
  const config = engine.randomConfigs[ui.selectedTrack]
  const minNote = config.pitch.low
  const maxNote = config.pitch.high
  const range = maxNote - minNote || 1

  // Step bars — 2 rows of 8
  for (let row = 0; row < 2; row++) {
    const barBaseY = STEP_AREA_TOP + (row + 1) * BAR_MAX_H + row * ROW_GAP
    for (let col = 0; col < COLS; col++) {
      const i = row * COLS + col
      const stepIdx = pageOffset + i
      const x = PAD + col * (STEP_W + COL_GAP)

      if (stepIdx >= track.pitch.length) {
        fillRect(ctx, { x, y: barBaseY - 2, w: STEP_W, h: 2 }, '#111118')
        continue
      }

      const pitchStep = track.pitch.steps[stepIdx]
      const normalized = Math.max(0, Math.min(1, (pitchStep.note - minNote) / range))
      const barH = Math.max(3, normalized * (BAR_MAX_H - 4))
      const barY = barBaseY - barH
      const isSelected = i === ui.selectedStep
      const color = isSelected ? COLORS.textBright : trackColor

      // Slide: tint bar orange when slide active
      const barColor = pitchStep.slide > 0 ? '#ff8800' : color

      fillRect(ctx, { x, y: barY, w: STEP_W, h: barH }, barColor)

      if (isSelected) {
        strokeRect(ctx, { x: x - 1, y: barY - 1, w: STEP_W + 2, h: barH + 2 }, '#ffffff', 1)
      }

      // Slide connector — diagonal line from this bar top to next bar top
      if (pitchStep.slide > 0) {
        const nextCol = col + 1
        const nextRow = nextCol >= COLS ? row + 1 : row
        const nextColWrapped = nextCol % COLS
        const nextStepIdx = pageOffset + (nextRow * COLS + nextColWrapped)
        if (nextRow < 2 && nextStepIdx < track.pitch.length) {
          const nextNorm = Math.max(0, Math.min(1, (track.pitch.steps[nextStepIdx].note - minNote) / range))
          const nextBarH = Math.max(3, nextNorm * (BAR_MAX_H - 4))
          const nextBarBaseY = STEP_AREA_TOP + (nextRow + 1) * BAR_MAX_H + nextRow * ROW_GAP
          const nextX = PAD + nextColWrapped * (STEP_W + COL_GAP)
          ctx.beginPath()
          ctx.moveTo(x + STEP_W, barY)
          ctx.lineTo(nextX, nextBarBaseY - nextBarH)
          ctx.strokeStyle = '#ff8800'
          ctx.lineWidth = 2
          ctx.stroke()
        }
      }

      // Playhead
      if (stepIdx === track.pitch.currentStep) {
        fillRect(ctx, { x, y: barBaseY + 2, w: STEP_W, h: 3 }, '#ffffff')
      }
    }
  }
}

File: crates/renderer/src/screens/pitch_edit.rs
use embedded_graphics::{pixelcolor::Rgb565, prelude::*};
use requencer_engine::types::SequencerState;

use crate::{colors, draw, layout, types::UiState};

/// Render the pitch edit screen: 2x8 grid showing MIDI note values.
pub fn render<D: DrawTarget<Color = Rgb565>>(
    display: &mut D,
    state: &SequencerState,
    ui: &UiState,
) {
    let track_idx = ui.selected_track as usize;
    let track = &state.tracks[track_idx];
    let color = colors::TRACK[track_idx];
    let dim = colors::TRACK_DIM[track_idx];
    let page_offset = ui.current_page as usize * 16;
    let (cell_w, cell_h) = layout::edit_step_size();

    // Header
    draw::fill_rect(
        display,
        0,
        layout::CONTENT_Y as i32,
        layout::LCD_W,
        layout::EDIT_HEADER_H,
        colors::STATUS_BAR,
    );
    draw::text(
        display,
        layout::PAD as i32,
        layout::CONTENT_Y as i32 + 4,
        "PITCH",
        color,
    );

    // Selected step info: note name + MIDI number + slide
    let sel = ui.selected_step as usize;
    if sel < track.pitch.steps.len() {
        let step = &track.pitch.steps[sel];
        let (name, oct) = colors::midi_note_name(step.note);
        let mut info_buf = [0u8; 16];
        let info = if step.slide > 0.0 {
            let slide_ms = (step.slide * 1000.0) as u16;
            draw::fmt_buf(
                &mut info_buf,
                format_args!("{}{} ({}) S:{}ms", name, oct, step.note, slide_ms),
            )
        } else {
            draw::fmt_buf(&mut info_buf, format_args!("{}{} ({})", name, oct, step.note))
        };
        draw::text_right(
            display,
            layout::LCD_W as i32 - layout::PAD as i32,
            layout::CONTENT_Y as i32 + 9,
            info,
            colors::TEXT_DIM,
        );
    }

    // Step grid
    let grid_y = layout::CONTENT_Y as i32 + layout::EDIT_HEADER_H as i32;

    for row in 0..2u32 {
        for col in 0..8u32 {
            let step_idx = page_offset + (row * 8 + col) as usize;
            if step_idx >= track.pitch.steps.len() {
                continue;
            }
            let step = &track.pitch.steps[step_idx];
            let x = layout::PAD as i32 + col as i32 * cell_w as i32;
            let y = grid_y + row as i32 * cell_h as i32;
            let is_playhead = track.pitch.current_step as usize == step_idx;
            let is_selected = ui.selected_step as usize == step_idx;
            let gate_on = step_idx < track.gate.steps.len()
                && (track.gate.steps[step_idx].on || track.gate.steps[step_idx].tie);

            // Cell background (trackDim tint like TS)
            let bg = if is_playhead {
                colors::SELECTED_ROW
            } else {
                dim
            };
            draw::fill_rect(display, x + 1, y + 1, cell_w - 2, cell_h - 2, bg);

            // Note bar (height proportional to MIDI note)
            let bar_color = if is_selected {
                colors::TEXT_BRIGHT
            } else if gate_on {
                color
            } else {
                dim
            };
            let bar_h = ((step.note as u32) * (cell_h - 16)) / 127;
            let bar_y = y + (cell_h as i32 - bar_h as i32 - 4);
            draw::fill_rect(display, x + 2, bar_y, cell_w - 4, bar_h.max(1), bar_color);

            // Note name text
            let (name, octave) = colors::midi_note_name(step.note);
            let mut buf = [0u8; 16];
            let note_str = draw::fmt_buf(&mut buf, format_args!("{}{}", name, octave));
            draw::text_center(
                display,
                x + cell_w as i32 / 2,
                y + 2,
                note_str,
                if gate_on { colors::TEXT } else { colors::TEXT_DIM },
            );

            // Slide indicator
            if step.slide > 0.0 {
                draw::fill_rect(display, x + 2, y + cell_h as i32 - 3, cell_w - 4, 2, colors::ACCENT);
            }

            // Playhead indicator
            if is_playhead {
                draw::playhead_bar(display, x, y, cell_w, cell_h);
            }

            if is_selected {
                draw::stroke_rect(display, x, y, cell_w, cell_h, colors::TEXT_BRIGHT);
            }
        }
    }

    // Footer
    let footer_y = layout::LCD_H as i32 - layout::EDIT_FOOTER_H as i32;
    draw::fill_rect(
        display,
        0,
        footer_y,
        layout::LCD_W,
        layout::EDIT_FOOTER_H,
        colors::STATUS_BAR,
    );
}
